### PR TITLE
[Merged by Bors] - chore(CategoryTheory): make `Functor.Full` a Prop

### PR DESCRIPTION
**Pull request by joelriou**
Before this PR, `Functor.Full` contained the data of the preimage of maps by a full functor `F`. This PR makes `Functor.Full` a proposition. This is to prevent any diamond to appear.

The lemma `Functor.image_preimage` is also renamed `Functor.map_preimage`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Functor/FullyFaithful.lean [hunk 1]
```diff
@@ -39,20 +37,12 @@ variable {C : Type u₁} [Category.{v₁} C] {D : Type u₂} [Category.{v₂} D]
...
 
 See <https://stacks.math.columbia.edu/tag/001C>.
 -/
-class Full (F : C ⥤ D) where
-  /-- The data of a preimage for every `f : F.obj X ⟶ F.obj Y`. -/
-  preimage : ∀ {X Y : C} (_ : F.obj X ⟶ F.obj Y), X ⟶ Y
-  /-- The property that `Full.preimage f` of maps to `f` via `F.map`. -/
-  witness : ∀ {X Y : C} (f : F.obj X ⟶ F.obj Y), F.map (preimage f) = f := by aesop_cat
+class Full (F : C ⥤ D) : Prop where
+  surjective {X Y : C} : Function.Surjective (F.map (X := X) (Y := Y))
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Functor/FullyFaithful.lean:44]: I think we should call this `map_surjective` to parallel `Functor.Faithful`.

Comment (joelriou) [reply Mathlib/CategoryTheory/Functor/FullyFaithful.lean]: Yes. I have just made the change.

File: Mathlib/Geometry/RingedSpace/SheafedSpace.lean [hunk 2]
```diff
@@ -189,7 +189,7 @@ set_option linter.uppercaseLean3 false in
 
 /-- The restriction of a sheafed space `X` to the top subspace is isomorphic to `X` itself.
 -/
-def restrictTopIso (X : SheafedSpace C) : X.restrict (Opens.openEmbedding ⊤) ≅ X :=
+noncomputable def restrictTopIso (X : SheafedSpace C) : X.restrict (Opens.openEmbedding ⊤) ≅ X :=
```
Comment (erdOne) [inline Mathlib/Geometry/RingedSpace/SheafedSpace.lean:192]: This gives bad defeqs. 
Maybe 
```
where 
hom := X.toPresheafedSpace.restrictTopIso.hom
inv := X.toPresheafedSpace.restrictTopIso.inv
```

File: Mathlib/CategoryTheory/Types.lean [hunk 3]
```diff
@@ -226,7 +226,7 @@ theorem uliftFunctor_map {X Y : Type u} (f : X ⟶ Y) (x : ULift.{v} X) :
 #align category_theory.ulift_functor_map CategoryTheory.uliftFunctor_map
 
 instance uliftFunctorFull : Functor.Full.{u} uliftFunctor where
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Types.lean:228]: This name needs to change.

File: Mathlib/CategoryTheory/Sites/SheafHom.lean [hunk 4]
```diff
@@ -218,10 +218,10 @@ def sheafHom' (F G : Sheaf J A) : Cᵒᵖ ⥤ Type _ where
     exact congr_fun ((presheafHom F.1 G.1).map_comp f g) φ.1
 
 /-- The canonical isomorphism `sheafHom' F G ≅ presheafHom F.1 G.1`. -/
-def sheafHom'Iso (F G : Sheaf J A) :
+noncomputable def sheafHom'Iso (F G : Sheaf J A) :
     sheafHom' F G ≅ presheafHom F.1 G.1 :=
   NatIso.ofComponents
-    (fun _ => Equiv.toIso (equivOfFullyFaithful (sheafToPresheaf _ _))) (fun _ => rfl)
+    (fun _ => Sheaf.homEquiv.toIso) (fun _ => rfl)
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Sites/SheafHom.lean:224]: Why is this noncomputable now?

File: Mathlib/CategoryTheory/Localization/Predicate.lean [hunk 5]
```diff
@@ -277,6 +276,14 @@ instance : (whiskeringLeftFunctor' L W E).Faithful := by
   infer_instance
   apply InducedCategory.faithful -- why is it not found automatically ???
 
+lemma full_whiskeringLeft : ((whiskeringLeft C D E).obj L).Full :=
+  (inferInstance : (whiskeringLeftFunctor' L W E).Full)
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Predicate.lean:280]:
```suggestion
  inferInstanceAs (whiskeringLeftFunctor' L W E).Full
```

File: Mathlib/CategoryTheory/Localization/Predicate.lean [hunk 6]
```diff
@@ -277,6 +276,14 @@ instance : (whiskeringLeftFunctor' L W E).Faithful := by
...
+
+lemma faithful_whiskeringLeft : ((whiskeringLeft C D E).obj L).Faithful :=
+  (inferInstance : (whiskeringLeftFunctor' L W E).Faithful)
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Predicate.lean:283]:
```suggestion
  inferInstanceAs (whiskeringLeftFunctor' L W E).Faithful
```

File: Mathlib/CategoryTheory/Functor/FullyFaithful.lean [hunk 7]
```diff
@@ -79,35 +69,23 @@ theorem mapIso_injective (F : C ⥤ D) [Faithful F] :
   Iso.ext (map_injective F (congr_arg Iso.hom h : _))
 #align category_theory.functor.map_iso_injective CategoryTheory.Functor.mapIso_injective
 
+theorem map_surjective (F : C ⥤ D) [Full F] :
+    Function.Surjective (F.map : (X ⟶ Y) → (F.obj X ⟶ F.obj Y)) :=
+  Full.map_surjective
+#align category_theory.functor.map_surjective CategoryTheory.Functor.map_surjective
+
 /-- The specified preimage of a morphism under a full functor. -/
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Functor/FullyFaithful.lean:77]:
```suggestion
/-- The choice of a preimage of a morphism under a full functor. -/
```

File: Mathlib/CategoryTheory/Functor/FullyFaithful.lean [hunk 8]
```diff
@@ -79,35 +69,23 @@ theorem mapIso_injective (F : C ⥤ D) [Faithful F] :
...
 @[pp_dot]
-def preimage (F : C ⥤ D) [Full F] (f : F.obj X ⟶ F.obj Y) : X ⟶ Y :=
-  Full.preimage.{v₁, v₂} f
+noncomputable def preimage (F : C ⥤ D) [Full F] (f : F.obj X ⟶ F.obj Y) : X ⟶ Y :=
+  (F.map_surjective f).choose
 #align category_theory.functor.preimage CategoryTheory.Functor.preimage
 
 @[simp]
 theorem image_preimage (F : C ⥤ D) [Full F] {X Y : C} (f : F.obj X ⟶ F.obj Y) :
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Functor/FullyFaithful.lean:84]: This should be called `map_preimage` instead...
Not for this PR though.

File: Mathlib/CategoryTheory/Category/Grpd.lean [hunk 9]
```diff
@@ -91,7 +91,7 @@ def forgetToCat : Grpd.{v, u} ⥤ Cat.{v, u} where
 set_option linter.uppercaseLean3 false in
 #align category_theory.Groupoid.forget_to_Cat CategoryTheory.Grpd.forgetToCat
 
-instance forgetToCatFull : forgetToCat.Full where preimage := id
+instance forgetToCatFull : forgetToCat.Full where map_surjective f := ⟨f, rfl⟩
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Category/Grpd.lean:94]: This name should be changed.

File: Mathlib/CategoryTheory/Adjunction/Opposites.lean [hunk 10]
```diff
@@ -214,7 +215,8 @@ theorem leftAdjointUniq_refl {F : C ⥤ D} {G : D ⥤ C} (adj1 : F ⊣ G) :
 #align category_theory.adjunction.left_adjoint_uniq_refl CategoryTheory.Adjunction.leftAdjointUniq_refl
 
 /-- If `G` and `G'` are both right adjoint to `F`, then they are naturally isomorphic. -/
-def rightAdjointUniq {F : C ⥤ D} {G G' : D ⥤ C} (adj1 : F ⊣ G) (adj2 : F ⊣ G') : G ≅ G' :=
+noncomputable def rightAdjointUniq {F : C ⥤ D} {G G' : D ⥤ C} (adj1 : F ⊣ G) (adj2 : F ⊣ G') :
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Adjunction/Opposites.lean:218]: Why is this now noncomputable?

File: Mathlib/Algebra/Homology/Single.lean [hunk 11]
```diff
@@ -118,10 +118,10 @@ instance (j : ι) : (single V c j).Faithful where
       ← single_map_f_self, w]
 
 noncomputable instance (j : ι) : (single V c j).Full where
```
Comment (erdOne) [inline Mathlib/Algebra/Homology/Single.lean:120]:
```suggestion
instance (j : ι) : (single V c j).Full where
```

File: Mathlib/Algebra/Category/MonCat/Basic.lean [hunk 12]
```diff
@@ -409,6 +409,7 @@ set_option linter.uppercaseLean3 false in
 -- automatically reflects isomorphisms
 -- we could have used `CategoryTheory.ConcreteCategory.ReflectsIso` alternatively
 @[to_additive]
-instance CommMonCat.forget₂Full : (forget₂ CommMonCat MonCat).Full where preimage f := f
+instance CommMonCat.forget₂Full : (forget₂ CommMonCat MonCat).Full where
```
Comment (erdOne) [inline Mathlib/Algebra/Category/MonCat/Basic.lean:412]: This name needs to be changed.

**Review by erdOne — COMMENTED** [review]
Thanks a lot!

File: Mathlib/CategoryTheory/Sites/SheafHom.lean (hunk 4 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/SheafHom.lean:224]: It became noncomputable when `equivOfFullyFaithful` became so, but when I used `Sheaf.homEquiv` to get better defeq, I forgot to remove the tag...

File: Mathlib/CategoryTheory/Functor/FullyFaithful.lean (hunk 8 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Functor/FullyFaithful.lean]: I had also spotted this bad name for a certain time... This is a good occasion to change it!

**Comment by joelriou** [comment]
Thanks very much @erdOne for the review!

File: Mathlib/CategoryTheory/Adjunction/Opposites.lean [hunk 13]
```diff
@@ -322,16 +324,16 @@ theorem rightAdjointUniq_refl {F : C ⥤ D} {G : D ⥤ C} (adj1 : F ⊣ G) :
 /-- Given two adjunctions, if the left adjoints are naturally isomorphic, then so are the right
 adjoints.
 -/
-def natIsoOfLeftAdjointNatIso {F F' : C ⥤ D} {G G' : D ⥤ C} (adj1 : F ⊣ G) (adj2 : F' ⊣ G')
-    (l : F ≅ F') : G ≅ G' :=
+noncomputable def natIsoOfLeftAdjointNatIso {F F' : C ⥤ D} {G G' : D ⥤ C}
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Adjunction/Opposites.lean:327]:
```suggestion
def natIsoOfLeftAdjointNatIso {F F' : C ⥤ D} {G G' : D ⥤ C}
```

File: Mathlib/CategoryTheory/Adjunction/Opposites.lean [hunk 14]
```diff
@@ -322,16 +324,16 @@ theorem rightAdjointUniq_refl {F : C ⥤ D} {G : D ⥤ C} (adj1 : F ⊣ G) :
...
+    (adj1 : F ⊣ G) (adj2 : F' ⊣ G') (l : F ≅ F') : G ≅ G' :=
   rightAdjointUniq adj1 (adj2.ofNatIsoLeft l.symm)
 #align category_theory.adjunction.nat_iso_of_left_adjoint_nat_iso CategoryTheory.Adjunction.natIsoOfLeftAdjointNatIso
 
 /-- Given two adjunctions, if the right adjoints are naturally isomorphic, then so are the left
 adjoints.
 -/
-def natIsoOfRightAdjointNatIso {F F' : C ⥤ D} {G G' : D ⥤ C} (adj1 : F ⊣ G) (adj2 : F' ⊣ G')
-    (r : G ≅ G') : F ≅ F' :=
+noncomputable def natIsoOfRightAdjointNatIso {F F' : C ⥤ D} {G G' : D ⥤ C}
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Adjunction/Opposites.lean:335]:
```suggestion
def natIsoOfRightAdjointNatIso {F F' : C ⥤ D} {G G' : D ⥤ C}
```

File: Mathlib/CategoryTheory/ConcreteCategory/BundledHom.lean [hunk 15]
```diff
@@ -152,7 +152,8 @@ instance forget₂ (F : ∀ {α}, d α → c α) [ParentProjection @F] :
 #align category_theory.bundled_hom.forget₂ CategoryTheory.BundledHom.forget₂
 
 instance forget₂Full (F : ∀ {α}, d α → c α) [ParentProjection @F] :
```
Comment (erdOne) [inline Mathlib/CategoryTheory/ConcreteCategory/BundledHom.lean:154]:
```suggestion
instance forget₂_full (F : ∀ {α}, d α → c α) [ParentProjection @F] :
```

**Comment by erdOne** [comment]
!bench

**Comment by erdOne** [comment]
Thanks a lot!
Maintainer merge

**Comment by erdOne** [comment]
maintainer merge please

**Comment by jcommelin** [comment]
What is the motivation for this PR? The speedup that the bench bot reported above? Or defeq problems? Or something else?

Please add some explanation to the commit message for this PR.

Otherwise, LGTM

bors d+

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**